test/e2e/apps/deployment.go

```diff
@@ -1705,3 +1709,146 @@ func testDeploymentSubresources(ctx context.Context, f *framework.Framework) {
+
+	rs := &appsv1.ReplicaSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-replicaset",
+			Namespace: ns,
+			Labels: map[string]string{
+				"test-name":                            "pod-template-hash-compatible",
+				appsv1.DefaultDeploymentUniqueLabelKey: "hash",
+			},
+		},
+		Spec: appsv1.ReplicaSetSpec{
+			Replicas: &size,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"test-name":                            "pod-template-hash-compatible",
+					appsv1.DefaultDeploymentUniqueLabelKey: "hash",
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"test-name":                            "pod-template-hash-compatible",
+						appsv1.DefaultDeploymentUniqueLabelKey: "hash",
+					},
+				},
+				Spec: v1.PodSpec{
+					TerminationGracePeriodSeconds: &zero,
+					Containers:                    []v1.Container{container},
+				},
+			},
+		},
+	}
+
+	_, err := c.AppsV1().ReplicaSets(ns).Create(ctx, rs, metav1.CreateOptions{})
+	framework.ExpectNoError(err, "Creating replica set %q in namespace %q", rs.Name, ns)
+
+	err = e2ereplicaset.WaitForReadyReplicaSet(ctx, c, ns, rs.Name)
+	framework.ExpectNoError(err)
+
+	// create the deployment to control the replicaset
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-deployment",
+			Namespace: ns,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &size,
+			Selector: &metav1.LabelSelector{MatchLabels: map[string]string{"test-name": "pod-template-hash-compatible"}},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"test-name": "pod-template-hash-compatible"},
+				},
+				Spec: v1.PodSpec{
+					TerminationGracePeriodSeconds: &zero,
+					Containers:                    []v1.Container{container},
+				},
+			},
+		},
+	}
+	deployment, err := c.AppsV1().Deployments(ns).Create(ctx, d, metav1.CreateOptions{})
+	framework.ExpectNoError(err, "Creating deployment %q in namespace %q", d.Name, ns)
+	err = e2edeployment.WaitForDeploymentComplete(c, deployment)
+	framework.ExpectNoError(err)
+
+	if err := checkPodTemplateHash(ctx, 1, true, false); err != nil {
+		framework.Failf("failed checking labels after create the deployment, %v", err)
+	}
+
+	// scale should be ok with old label
+	newSize := int32(2)
+	d.Spec.Replicas = &newSize
+	deployment, err = c.AppsV1().Deployments(ns).Update(ctx, d, metav1.UpdateOptions{})
+	framework.ExpectNoError(err, "scale up deployment %q in namespace %q", d.Name, ns)
+	err = e2edeployment.WaitForDeploymentComplete(c, deployment)
+	framework.ExpectNoError(err)
+	if err := checkPodTemplateHash(ctx, 2, true, false); err != nil {
+		framework.Failf("failed checking labels after scale up the deployment, %v", err)
+	}
+
+	// update the deployment to create new labels
+	d.Spec.Replicas = &size
+	d.Spec.Template.Spec.Containers[0].Image = imageutils.GetE2EImage(imageutils.Httpd)
```

Comment:
it seems unnecessary for this test to depend on 2 images. Can we for example update an env variable instead?